### PR TITLE
Update of Power Management functions

### PR DESCRIPTION
**Pull request by Karlson2k**
For all platforms: 
- put more information to the log in PowerManager (currently it log minimal information, sometimes it's hard to track processes before Suspend/Shutdown), 
- prevent possible crashes if PowerManager functions called Initialize()

For win32:
- prevent lock of main thread on Suspend/Hibernate
- add more error and debug logging
- move all power management functions from Win32Utils to Win32PowerSyscall

### CODE REVIEW

**Comment by topfs2** [comment]
I can't say anything about the win32 stuff but the general alterations in power manager looks alright to me. I kindof dislike the extra tracking of the state (or state tracking in general :P) but perhaps thats the job of the power manager no matter :) And its nice if addons etc doesn't get double notifications. Is it just win32 that has that problem? Because I think the others doesn't publish multiple sleeps?

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 1]
```diff
@@ -18,14 +18,84 @@
...
 bool CWin32PowerSyscall::m_OnResume = false;
 bool CWin32PowerSyscall::m_OnSuspend = false;
+bool CWin32PowerSyscall::m_sleeping = false;
+
+
+// local helpers
+static DWORD WINAPI threadForHibernateOs(_In_ LPVOID ignored)
+{
+  CLog::LogF(LOGNOTICE, "Requesting OS to hibernate");
+  if (SetSuspendState(true, true, false) != FALSE) // may be blocked until resume
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: should use TRUE/FALSE and not true/false

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 2]
```diff
@@ -18,14 +18,84 @@
...
+    return TRUE;
+  }
+  CLog::LogF(LOGERROR, "Can't hibernate system, error code: %lu", GetLastError());
+  return FALSE;
+}
+
+static DWORD WINAPI threadForSuspendOs(_In_ LPVOID ignored)
+{
+  CLog::LogF(LOGNOTICE, "Requesting OS to suspend");
+  if (SetSuspendState(false, true, false) != FALSE) // may be blocked until resume
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: Should use TRUE/FALSE and not true/false

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 3]
```diff
@@ -33,41 +103,142 @@ CWin32PowerSyscall::CWin32PowerSyscall()
...
 bool CWin32PowerSyscall::Powerdown()
 {
-  return CWIN32Util::PowerManagement(POWERSTATE_SHUTDOWN);
+  if (!AdjustPrivileges())
+    return false;
+
+  CLog::LogF(LOGINFO, "Requesting OS shutdown");
+
+  if (g_sysinfo.IsWindowsVersionAtLeast(CSysInfo::WindowsVersionWin8))
+    return InitiateShutdownW(NULL, NULL, 0, SHUTDOWN_HYBRID | SHUTDOWN_INSTALL_UPDATES | SHUTDOWN_POWEROFF,
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: Should compare to FALSE and not rely on implicit conversion from BOOL to bool. This goes for the other lines as well

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 4]
```diff
@@ -75,22 +246,49 @@ bool CWin32PowerSyscall::CanReboot()
 
 int CWin32PowerSyscall::BatteryLevel()
 {
-  return CWIN32Util::BatteryLevel();
+  SYSTEM_POWER_STATUS SystemPowerStatus;
+
+  if (GetSystemPowerStatus(&SystemPowerStatus) && SystemPowerStatus.BatteryLifePercent != 255)
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: maybe stick a comment in here that 255 means unable to read battery value?

**Comment by Paxxi** [comment]
Overall this looks good. I do Think however that if a thread fails to create the action shouldn't just be ignored but fall back to blocking. It isn't likely that the user will mind during a suspend/reboot anyway

**Comment by MartijnKaijser** [comment]
If still needed please open a new PR and reference this one.

**Pull request closed without merging**